storybookjs/storybook
# Addon-docs: support vue inline rendering

Issue: #6706 

## What I did

- Added a parameter to `docs` configuration object called `prepareForInline`
- Added dependency [@egoist/vue-to-react](https://github.com/egoist/vue-to-react) for the inline transform
- Added a vue -> react transformation to the `vue/config.js` file in the `addon-docs` folder
- ~~Removed explicit heights from the MDX `<Story />` blocks in the vue-kitchen-sink example~~ (Put these back, since inline stories will be opt-in for now)
- Added a first pass at docs for inline parameters
- Added an inline example to Vue kitchen sink

## How to test

- Check that the inline story in Vue Kitchen sink behaves as expected.
- Add `inlineStories: true` to the vue kitchen sink config, build it and make sure everything looks the way you expect
- Add `inlineStories: true` to the angular kitchen sink config and ensure the appropriate error is thrown
- Read Documentation additions and edit/critique/remove as you see fit 👍 

## Discussion points:

- This kinda breaks the `centered` add-on in docs mode. I tried to disable it, but since it's implemented as a decorator, I couldn't figure out how to use the `match` parameter the same way that @shilman is using to disable the background addon.
- When an explicit height was set on the the `<Story>` block it seemed to cause some overflow issues. Removing the height made everything work fine, but maybe some `overflow` styles should be added in case the user _wants_ to set an explicit height for some reason? I was assuming that was a cross-framework issue, so I didn't worry about addressing the issue here.

## FAQ (A.k.a questions I frequently asked myself)

### Why not just use a decorator? Why the new docs config property?
This could totally be my lack of knowledge of the codebase, but I couldn't figure out any way to _guarantee_ that my decorator would be the _last one_ to run. And after the transform to React land runs, the ship has sailed. None of the other decorators can run at that point because we're dealing with a react element now. The most straightforward way, from my perspective, was to add a new config that allows you to specify _how_ a story should be prepared to be inlined. Or how it should be cleanly introduced to the React environment. If any of the core maintainers think a decorator would be cleaner/more consistent and there's a way I can easily guarantee my vue-to-react decorator is the last thing in the decorator pipeline to run, I'd be more than glad to pivot on that approach.

Oh, and last but not least, decorators registered to the vue storybook have a type signature of `DecoratorFunction<VueConstructor>` which, by design, _is not_ what the inline transformer returns.

### Why not `vuera` (or some other library)? It has so many 🌟 's
Vuera is pretty robust, heavy, and handles a bunch of edge cases that I don't think we really care about all that much in our use case. `@egoist/vue-to-react`, on the other hand, is [dead-simple react hook single-file implementation](https://github.com/egoist/vue-to-react/blob/master/src/index.js) that should do everything we need it to do. Egoist is an incredibly talented open source developer who has worked in the Vue community for a long time now, and he wrote this tiny library for our _exact_ use-case, to convert Vue components to be renderable in his custom component [demo tool](https://github.com/egoist/demoboard).

Looking forward to comments and criticism 👍
